source/mainnet/smart-contracts/guides/upgradeable-contract.rst

```diff
@@ -0,0 +1,67 @@
+.. _guide-upgradable-contract:
+
+===========================
+Make a contract upgradeable
+===========================
+
+This guide will show how to make a Rust smart contract upgradeable.
+
+Immutable smart contracts comes with the drawback that bugs cannot be fixed, new features and cost optimizations cannot be implemented.
+For some decentralized applications this is a problem and these need some way to upgrade the smart contract code.
+
+.. note::
+   A common way to workaround immutability, seen on other blockchains, is to have an additinal smart contract acting as a proxy, forwarding calls to the actual smart contract.
+   Upgrading is then updating the proxy to forward the calls to another smart contract.
+   Unfortunately, this will introduce an extra cost on every contract call and add an additional layer of complexity to the contract code.
+
+Concordium smart contract instances have the option to upgrade their smart contract module by calling the ``upgrade`` function with a reference to a new smart contract module to use.
+If successful, new invocations to the upgraded smart contract instance will use the new smart contract module.
+
+.. seealso::
+
+   To learn more about upgradeability of smart contracts see :ref:`contract-instance-upgradeability`.
+
+Owner can upgrade
+=================
+
+The code below will add an ``upgrade`` endpoint to a smart contract, which allows the smart contract owner to trigger a smart contract upgrade.
+The parameter for this endpoint takes the new module reference and optionally a name of an entrypoint with parameter to invoke in the upgraded smart contract instance.
+
+Providing the optional entrypoint can be used for triggering a migration function in the new module.
+This has the benefit of being in the same transaction as the upgrade itself, making the transaction reject reverting the upgrade if the migration failed.
+
+.. code-block:: rust
+
+   /// The parameter type for the contract function `upgrade`.
+   /// Takes the new module and optionally a migration function to call in the new
+   /// module after the upgrade.
+   #[derive(Debug, Serialize)]
+   struct UpgradeParams {
+       /// The new module reference.
+       module:  ModuleReference,
+       /// Migration function in the new module.
+       migrate: Option<(OwnedEntrypointName, Vec<u8>)>,
```

Comment:
Don't you want to use the `OwnedParameter` type here?

Reply:
Or is that not avialable in concordium-std?

Reply:
it is available